pkg/keystorage/recstore/recstore.go

```diff
@@ -0,0 +1,171 @@
+	}
+
+	data, err := s.unmarshal(resp.Kvs[0].Value)
+	if err != nil {
+		var zero T
+
+		return Result[T]{}, fmt.Errorf("error unmarshaling data for '%T': %w", zero, err)
+	}
+
+	return Result[T]{
+		Res:     data,
+		Version: resp.Kvs[0].Version,
+	}, nil
+}
+
+// Create marshals and stores the value for the given key for the selected version.
+// It returns an error if the key already exists.
+func (s *RecStore[T]) Create(ctx context.Context, val T) error {
+	etcdKey := s.etcdKey()
+
+	marshal, err := s.marshal(val)
+	if err != nil {
+		return fmt.Errorf("error marshaling data for '%T': %w", val, err)
+	}
+
+	txnResp, err := s.client.Txn(ctx).If(
+		clientv3.Compare(clientv3.Version(etcdKey), "=", 0), // not exists check
+	).Then(
+		clientv3.OpPut(etcdKey, string(marshal)),
+		clientv3.OpGet(etcdKey),
+	).Commit()
+	if err != nil {
+		return fmt.Errorf("error creating key %s: %w", etcdKey, err)
+	}
+
+	if !txnResp.Succeeded {
+		return util.NewTypedError[AlreadyExists](fmt.Errorf("key %s already exists", etcdKey))
+	}
+
+	return nil
+}
+
+// Update marshals and stores the value for the given key for the selected version.
+// If version is 0, it will create a new record.
+// If version is not 0, it will update the record for the selected version.
+func (s *RecStore[T]) Update(ctx context.Context, val T, version int64) error {
+	if version == 0 {
+		return s.Create(ctx, val)
+	}
+
+	ctx = util.ClearContextMeta(ctx)
+
+	etcdKey := s.etcdKey()
+
+	getResp, err := s.client.Get(ctx, etcdKey)
+	if err != nil {
+		return fmt.Errorf("error getting key %s: %w", etcdKey, err)
+	}
+
+	if len(getResp.Kvs) == 0 {
+		return util.NewTypedError[NotFound](fmt.Errorf("key %s not found", etcdKey))
+	}
+
+	actualVersion := getResp.Kvs[0].Version
+	if version != actualVersion {
+		return util.NewTypedError[VersionConflict](fmt.Errorf("key %s version mismatch, expected '%d' actual '%d'", etcdKey, version, actualVersion))
+	}
+
+	marshal, err := s.marshal(val)
+	if err != nil {
+		return fmt.Errorf("error marshaling data for '%T': %w", val, err)
+	}
+
+	txnResp, err := s.client.Txn(ctx).If(
+		clientv3.Compare(clientv3.Version(etcdKey), "=", version),
+	).Then(
+		clientv3.OpPut(etcdKey, string(marshal)),
+		clientv3.OpGet(etcdKey),
+	).Else(
+		clientv3.OpGet(etcdKey),
+	).Commit()
+	if err != nil {
+		return fmt.Errorf("error updating key %s: %w", etcdKey, err)
+	}
+
+	if !txnResp.Succeeded {
+		var foundVersion int64
+
+		txnGetKvs := txnResp.Responses[0].GetResponseRange().Kvs
+		if len(txnResp.Responses[0].GetResponseRange().Kvs) > 0 {
+			foundVersion = txnGetKvs[0].Version
+		}
+
+		return util.NewTypedError[TxVersionConflict](fmt.Errorf("key %s tx version mismatch, expected '%d' actual '%d'", etcdKey, version, foundVersion))
+	}
+
+	return nil
+}
+
+// Result is a result of [RecStore.Get].
+type Result[T any] struct {
+	Res     T
+	Version int64
+}
+
+func (s *RecStore[T]) etcdKey() string {
+	return "record-store-" + s.prefix
```

Comment:
today COSI resources are stored like this:

```
/<prefix>/<namespace>/<type>/<hashed_id>
```

so we should better do same thing here,

```
s.prefix + "record-store/" + name
``` ?

where `s.prefix` will be set to `/<instance-name>/` in the project.